graph_analysis/graph_feature_extraction.py

```diff
@@ -0,0 +1,104 @@
+"""This module extracts features in pandas DataFrame format for creating graph structure charts."""
+
+import networkx as nx
+import pandas as pd
+from scipy import stats
+import numpy as np
+
+
+class GraphFeatureExtraction:
+    """
+        A class to extract features from a networkx graph for A->B analysis.
+
+        ...
+
+        Methods:
+            get_cycle_stats()
+                Finds all cycles in a graph and returns stats and dataframe with node count in graph.
+
+            remove_outliers(df, distributions_max=4)
+                Removes outliers from distribution that are more than n standard deviations away from the mean.
+
+            node_distribution_stats()
+                Generates dataframes with paths from A->B for a line chart.
+    """
+    def __init__(self, G):
+        self.G = G
+        self.node_types = ["dataset", "system", "system_collection", "dataset_collection",
+                           "data_integrity", "processing", "collection"]
+        cycle_stats = self.get_cycle_stats()
+        self.number_of_cycles = cycle_stats["number_of_cycles"]
+        self.number_unique_nodes_in_cycles = cycle_stats["number_unique_nodes_in_cycles"]
+        self.df_cycle_count = cycle_stats["df_cycle_count"]
+
+        self.df_without_outliers, self.collection_outliers = self.node_distribution_stats()
+
+    def get_cycle_stats(self):
+        """Finds all cycles in a graph and returns stats and dataframe with node count in graph."""
+        cycles = list(nx.simple_cycles(self.G))
+        number_of_cycles = len(cycles)
+        length_of_cycles = [len(cycle) for cycle in cycles]
+        nodes_in_cycles = {node_type: set() for node_type in self.node_types}
+
+        for cycle in cycles:
+            for node in cycle:
+                nodes_in_cycles[node.split("_")[0]].add(node)
```

Comment:
Is this getting the id of the node because of its name "dataset_xxx"? What if I have nodes named differently?